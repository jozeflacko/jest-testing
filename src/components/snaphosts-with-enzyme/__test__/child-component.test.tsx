import * as Enzyme from 'enzyme';
import * as React from 'react';
import Child, { IProps, IState, IOnClicked } from '../child-component';

describe('<Child />', () => {

    const label = "My Label";
    const attrMsg = "Attr msg";
    const onClickedMock: IOnClicked = jest.fn((times:number)=>{});
    const props: IProps = {
        buttonLabel: label,
        onClicked: onClickedMock,
        attrMsg: attrMsg
    };
    const component = Enzyme.shallow(<Child {...props} />);

    it('renders correctly component', () => {      
        expect(component).toMatchSnapshot();
    });
    /* 
    it('has expected props',()=>{
        expect(component.prop('buttonLabel')).toEqual(label);
        expect(component.props().onClickedMock).toBe(onClickedMock);
    });
    */
    
    it('has expected state howManyTimesWasButtonClicked to be 0', ()=>{ 
        expect((component.state() as IState).howManyTimesWasButtonClicked).toBe(0);
    });

    describe('Contains expected button',()=>{

        const button = component.find('button.button');

        it('has correct label __buttonLabel__',()=>{
            expect(button.text()).toBe(label);
        });

        it('with expected class',()=>{
            expect(button.hasClass('button')).toBeTruthy();
        });
        it('contains passed attrMsg in a data-attrMsg attribute',()=>{
            const attr = (button.prop('data-attrMsg'));
            expect(attr).toEqual(attrMsg);
        });

        describe('try to click on a button',()=> {
            beforeEach(()=>{
                button.simulate('click');
            });
            it('after 1 click passed function should been called',()=>{
                
                // function was called
                expect(onClickedMock).toHaveBeenCalledTimes(1);
                // state was updated: howManyTimesWasButtonClicked from 0 will change to 1
                expect((component.state() as IState).howManyTimesWasButtonClicked).toEqual(1);
            });
        });

    });
});