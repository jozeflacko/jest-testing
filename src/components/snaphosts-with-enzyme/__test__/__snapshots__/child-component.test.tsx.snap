// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Child /> renders correctly component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Child
    attrMsg="Attr msg"
    buttonLabel="My Label"
    onClicked={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Main Component
        </h1>,
        <button
          className="button"
          data-attrMsg="Attr msg"
          onClick={[Function]}
        >
          My Label
        </button>,
      ],
      "className": "child-component",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Main Component",
        },
        "ref": null,
        "rendered": "Main Component",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "My Label",
          "className": "button",
          "data-attrMsg": "Attr msg",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "My Label",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Main Component
          </h1>,
          <button
            className="button"
            data-attrMsg="Attr msg"
            onClick={[Function]}
          >
            My Label
          </button>,
        ],
        "className": "child-component",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Main Component",
          },
          "ref": null,
          "rendered": "Main Component",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "My Label",
            "className": "button",
            "data-attrMsg": "Attr msg",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "My Label",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
